File: frontend/src/types/highlight.ts
export type HighlightColor = "yellow" | "green" | "blue" | "pink";

export interface HighlightPosition {
  start_offset?: number;
  end_offset?: number;
  page?: number; // PDF only
}

export interface Highlight {
  id: string;
  source_material_id: string;
  position: HighlightPosition;
  selected_text: string;
  color: HighlightColor;
  note: string | null;
  created_at: string;
  updated_at: string;
}

export interface HighlightCreate {
  position: HighlightPosition;
  selected_text: string;
  color?: HighlightColor;
  note?: string;
}

export interface HighlightUpdate {
  color?: HighlightColor;
  note?: string;
}

export interface TextSelection {
  text: string;
  startOffset: number;
  endOffset: number;
  page?: number; // PDF only
}

File: backend/app/models/highlight.py
from sqlalchemy import Column, String, Text, DateTime, ForeignKey, Enum, JSON
from sqlalchemy.orm import relationship
from datetime import datetime, timezone
import uuid
import enum

from app.database import Base


class HighlightColor(str, enum.Enum):
    YELLOW = "yellow"
    GREEN = "green"
    BLUE = "blue"
    PINK = "pink"


class Highlight(Base):
    __tablename__ = "highlights"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    source_material_id = Column(String, ForeignKey("source_materials.id", ondelete="CASCADE"), nullable=False, index=True)

    # Position tracking (JSON for flexibility across file types)
    # Text/MD: {"start_offset": int, "end_offset": int}
    # PDF: {"page": int, "start_offset": int, "end_offset": int}
    # DOCX: {"start_offset": int, "end_offset": int} (in converted HTML)
    position = Column(JSON, nullable=False)

    # The actual highlighted text (stored for display even if position changes)
    selected_text = Column(Text, nullable=False)

    # Highlight properties
    color = Column(Enum(HighlightColor), default=HighlightColor.YELLOW, nullable=False)

    # Optional note content
    note = Column(Text, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    # Relationships
    source_material = relationship("SourceMaterial", back_populates="highlights")
